[DAPHNE-#658] typeOf() function to determine the type of an object

Garic152 opened the pull request:
This PR implements https://github.com/daphne-eu/daphne/issues/357 and allows the user to determine the type of a Scalar or Object in DAPHNE DSL-Scrips using the typeOf() built-in function.

Features:
- `typeOf(scalar)`: Returns the C++ representation of the scalar value type using ValueTypeUtils.h.
- `typeOf(object)`: Returns the dimension and ValueType of an object as `(dim, scalarType)`.


Changes:
- Added the typeOf() DaphneDSL built-in function, the typeOf DaphneIR operation, as well as the typeOf() kernel
- Added corresponding script-level tests as a new test file TypeOfTest.cpp in `.../cli/operations`, ensuring that the correct types get printed for scalar types and both sparse and dense matrices

AlexRTer reviewed (changes requested):
Thank you for this contribution @Garic152!
A `typeOf` function is a very useful tool, especially for debugging DaphneDSL scripts or when exploring and verifying type conversions. The implementation **looks very good** to me already. I only made some minor changes such as changing the `TypeOfObj` kernel slightly to avoid having to handle the memory `scaType` points to.

In fact, this PR already highlighted a bug in the way the `DaphneDSLVisitor` handles parsing for special values of float type (e.g. `typeOf(inff)` is a double, not a float)!

There are **two small required changes** that still need to be made before merging.
1.  Please **add the `typeOf` function to the Builtins section of the documentation** (doc/DaphneDSL/Builtins.md) as well as a small description
2. **Add support for the output of one value type per column for Frames** (see the code comment for details)

Please let me know if you have any questions or would like to share your opinion on any changes I made.

Garic152 commented:
Thank you for reviewing the code, I totally overlooked the fact that the frames were able to contain multiple value types.
I changed the `typeOf(frame)` kernel to now iterate over the columns of the frame and return the result formatted like the `print()` function.
I also adjusted the existing tests and added another line testing the newly added functionality with all supported types in one frame.

Let me know if there are still some things missing and I'll add them as soon as possible!

AlexRTer commented:
Thank you for implementing all requested changes and in such a short time frame as well.
From my point of view, **this PR is ready to be merged.**

pdamme reviewed (approved):
Thanks for contributing this feature, @Garic152. Thanks for your helpful review and improvements, @AlexRTer. This PR **looks very good** to me, too. It's **ready to be merged**.

I slightly adjusted the positioning of the new `typeOf`-builtin/op by putting it into the "matrix dimensions" category of operations, which I renamed to "matrix meta data". Furthermore, I added a few more instantiations of the `typeOfSca`/`typeOfObj`-kernels, such that we can use them with any scalar or data object.
